src/hotspot/share/opto/vectornode.cpp

```diff
@@ -1879,6 +1879,39 @@ Node* ReverseVNode::Identity(PhaseGVN* phase) {
   return reverse_operations_identity(this, in(1));
 }
 
+// Optimize away redundant AndV/OrV nodes when the operation
+// is applied on the same input node multiple times
+static Node* redundant_logical_identity(Node* n) {
+  Node* n1 = n->in(1);
+  // (OperationV (OperationV src1 src2) src1) => (OperationV src1 src2)
+  // (OperationV (OperationV src1 src2) src2) => (OperationV src1 src2)
+  // (OperationV (OperationV src1 src2 m1) src1 m1) => (OperationV src1 src2 m1)
+  // (OperationV (OperationV src1 src2 m1) src2 m1) => (OperationV src1 src2 m1)
+  if (n->Opcode() == n1->Opcode()) {
+    if (((!n->is_predicated_vector() && !n1->is_predicated_vector()) ||
+       (n->is_predicated_vector() && n1->is_predicated_vector() &&
+       n->in(3) == n1->in(3))) && (n->in(2) == n1->in(1) || n->in(2) == n1->in(2))) {
```

Comment:
This condition is hard to parse, I would suggest:

```suggestion
    if (((!n->is_predicated_vector() && !n1->is_predicated_vector()) ||
         ( n->is_predicated_vector() &&  n1->is_predicated_vector() && n->in(3) == n1->in(3))) &&
        (n->in(2) == n1->in(1) || n->in(2) == n1->in(2))) {
```